fix incorrect Clipboard handling

xomx opened the pull request:
Fix #15324 , #15271 , #3054 ; possibly #9426 (for now, I would not close this issue, more below).

This PR mainly fixes some missing error-state checks for the OpenClipboard WINAPI used. It also enhances the behavior of the Notepad++ Clipboard History panel during the encoding conversions (now it hides some Scintilla Clipboard op used for these conversions).

Now I have no more time for an exact proof about fixing also the long standing #9426. My idea behind that statement is that a N++ Clipboard op (e.g. a paste-op) can somehow be influenced by a previous incorrect Clipboard manipulation/state, leading to some Scintilla malfunction (this was the exact case in the fixed #15324). The previous Notepad++ OpenClipboard WINAPI implementation was at many places vulnerable to a possible external Clipboard concurrent manipulation, so maybe the users influenced in the #9426 may also have some heavy using Clipboard SW (a monitoring tool) installed, which may increase the probability of such a conflict with the Notepad++. There is some coincidence in what a user says in the https://github.com/notepad-plus-plus/notepad-plus-plus/issues/9426#issuecomment-2167368080 : "the issue disappears after I restart Notepad++", which was also the exact case in the https://github.com/notepad-plus-plus/notepad-plus-plus/issues/15324#issuecomment-2181850779 . We will see - if there are no more such reports, we can take the fixing of #9426 as confirmed.

For future enhancement I also recommend to use an equivalent of the [ScintillaWin.cxx](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/6b441b738a951ed225632f4085a46c95cd4aa13f/scintilla/win32/ScintillaWin.cxx#L2798-L2808) func OpenClipboardRetry.

donho commented on PowerEditor/src/ScintillaComponent/ScintillaEditView.cpp:
```diff
@@ -4591,66 +4591,76 @@ bool ScintillaEditView::pasteToMultiSelection() const
-				}
-				execute(SCI_ENDUNDOACTION);
-				return true;
+				::CloseClipboard();
```
Halt operation.

donho commented on PowerEditor/src/ScintillaComponent/UserDefineDialog.cpp:
```diff
@@ -1734,21 +1734,22 @@ bool StringDlg::isAllowed([[maybe_unused]] const generic_string & txt)
-		HANDLE hClipboardData = GetClipboardData(CF_UNICODETEXT);
-		if (NULL != hClipboardData)
+		HANDLE hClipboardData = ::GetClipboardData(CF_UNICODETEXT);
+		if (hClipboardData)
```
Please use: `if (!hClipboardData) { ::CloseClipboard(); return;}`

donho reviewed (changes requested):
For the sake of readibility of code, please reduce the level of depth while using `if`.

xomx commented on PowerEditor/src/NppCommands.cpp:
```diff
@@ -3122,44 +3141,95 @@ void Notepad_plus::command(int id)
+							}
+							::GlobalUnlock(hClipboardData);
+						}
+					}
```
@donho 
>There are many levels of depth here. That makes hard to read the code.
OTOH, if hglbClipboardCopy is NULL, the operation should be halted.

I understand the request (I tried to follow exactly this in other places of the patch) but in this case it is valid to continue in execution of this part of code, even if one cannot store the previous Clipboard content to the hglbClipboardCopy. This will e.g. happen if the previous Clipboard data is so large (or the free memory is so much fragmented) that the GlobalAlloc fails (but it does not automatically mean that the following encoding conversion steps using the Clipboard will fail too).

I added that more indented code part there also because it is relatively short and clear...

donho replied:
@xomx 
By reading again my code (which has been done since long time ago), I do remember the logic now.
You're right about it!

xomx commented on PowerEditor/src/NppCommands.cpp:
```diff
@@ -3122,44 +3141,95 @@ void Notepad_plus::command(int id)
+						::EmptyClipboard();
+						::CloseClipboard();
+					}
+				}
```
@donho 
At this place it is perfectly legal when the hglbClipboardCopy is NULL (no previous Clipboard data to restore...).

xomx commented on PowerEditor/src/ScintillaComponent/ScintillaEditView.cpp:
```diff
@@ -4591,66 +4591,76 @@ bool ScintillaEditView::pasteToMultiSelection() const
-			size_t nbClipboardStr = clipboardStrings.size();
-
-			if (nbSelections >= nbClipboardStr) // enough holes for every insertion, keep holes empty if there are some left
+		}
```
Will do here.
Because this PR has started to be bigger, I tried to change as little code as possible. But this place really needs the changes you proposed.

xomx commented:
@donho 
When I debugged different encoding conversion scenarios here, I have found another problem (not directly connected to this issue) - when I converted some text containing also non-ASCII characters from ANSI to e.g. UTF-8, the cursor position could be incorrectly restored after such conversion. This is because the N++ does not shows real character count at `Pos:` (and `Length:`) but rather the text byte-count. So when I converted these 20 characters `0123456789ěščřžýáíéň` from the ANSI to UTF-8 with the cursor at its end (N++ showed `length: 20` and `Pos: 21`), the cursor has been incorrectly restored after the `ž` (N++ showed `length: 30` and `Pos: 21`).

So IMO the `Pos:` (and/or `Length:`) info should be either recalculated to real char-count or for the cursor restoration placement should be used the `Ln:` and `Col:` info instead of the `Pos:`.

donho commented:
@xomx 
> So IMO the Length: info should be recalculated to real char-count.

It's arguable. But "Length:" on status bar for me should be the file length rather the character count, so user opens a file he/she knows the file length. If user needs to know the character count, he/she can select all the document by hitting Ctrl-A, then "Sel:" will show the character count.

xomx commented:
@donho 
Ok. 
But the possible invalid cursor positioning after the conversion should be fixed in another issue/PR.

donho commented:
@xomx 
> For future enhancement I also recommend to use an equivalent of the [ScintillaWin.cxx](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/6b441b738a951ed225632f4085a46c95cd4aa13f/scintilla/win32/ScintillaWin.cxx#L2798-L2808) func OpenClipboardRetry (instead of the pure OpenClipboard).

OK for me - please do it in another PR - if you want, of course.
